src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,231 @@
+            bool recursive = false)
+        {
+            _originalUserPath = directory;
+            _originalFullPath = Path.GetFullPath(directory);
+            _recursive = recursive;
+            _predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            _transform = transform ?? throw new ArgumentNullException(nameof(transform));
+            _threadId = Environment.CurrentManagedThreadId;
+            _state = state;
+        }
+
+        private FindEnumerable(
+            string originalUserPath,
+            string originalFullPath,
+            FindTransform<TResult> transform,
+            FindPredicate<TState> predicate,
+            TState state,
+            bool recursive)
+        {
+            _originalUserPath = originalUserPath;
+            _originalFullPath = originalFullPath;
+            _predicate = predicate;
+            _transform = transform;
+            _state = state;
+            _recursive = recursive;
+            _threadId = Environment.CurrentManagedThreadId;
+        }
+
+        /// <summary>
+        /// Simple wrapper to allow creating a file handle for an existing directory.
+        /// </summary>
+        public static IntPtr CreateDirectoryHandle(string path)
+        {
+            IntPtr handle = Interop.Kernel32.CreateFile_IntPtr(
+                path,
+                Interop.Kernel32.FileOperations.FILE_LIST_DIRECTORY,
+                FileShare.ReadWrite | FileShare.Delete,
+                FileMode.Open,
+                Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle == IntPtr.Zero || handle == (IntPtr)(-1))
+            {
+                // Historically we throw directory not found rather than file not found
+                int error = Marshal.GetLastWin32Error();
+                if (error == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                    error = Interop.Errors.ERROR_PATH_NOT_FOUND;
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, path);
+            }
+
+            return handle;
+        }
+
+        public IEnumerator<TResult> GetEnumerator()
+        {
+            if (Interlocked.Exchange(ref _enumeratorCreated, 1) == 0 && _threadId == Environment.CurrentManagedThreadId)
+            {
+                InitEnumeration();
+                return this;
+            }
+
+            FindEnumerable<TResult, TState> clone = new FindEnumerable<TResult, TState>(_originalUserPath, _originalFullPath, _transform, _predicate, _state, _recursive);
+            clone.InitEnumeration();
+            return clone;
+        }
+
+        private void InitEnumeration()
+        {
+            // We want to delay creating any handles until the enumerator is actually created to
+            // avoid keeping handles open any longer than needed. Once we actually start enumerating
+            // it's fair game to stash handles until we've finished processing the given directory.
```

Comment:
Is this a behavior change?

On .NET Core 2.0, the following example throws `DirectoryNotFoundException`:

```c#
using System.IO;

class Program
{
    static void Main()
    {
        var enumerator = Directory.EnumerateFiles(@"C:\does\not\exist");
    }
}
```

Is that still the case after these changes? Or does it now throw when `GetEnumerator` is called?

Reply:
Yes, it is a behavior change. This particular case I think I'll try to replicate- giving it a bit more thought.